### ISSUE TITLE
`conan graph build-order` KeyError when using build_requires

### ISSUE THREAD

**Issue by sztomi** [conan-io/conan#5807]
Conan version: 1.18.5

When using `graph build-order` in a graph where some dependencies are build_requires, I'm getting a KeyError unhandled exception. This is most likely another facet of https://github.com/conan-io/conan/issues/5682 but I'm not able to work around it the same way I did with that one. I don't entirely understand what's going on, but managed to throw together a repo that reliably reproduces it:

https://github.com/sztomi/build_requires_keyerror

(pretty similar to the others I reported lately, I hope these do help)

This is the traceback:

```
Traceback (most recent call last):
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/command.py", line 1859, in run
    method(args[0][1:])
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/command.py", line 1748, in graph
    build_order = self._conan.build_order(args.lockfile, args.build)
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/conan_api.py", line 78, in wrapper
    return f(*args, **kwargs)
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/conan_api.py", line 1247, in build_order
    False, False, remotes, recorder)
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/graph/graph_manager.py", line 182, in load_graph
    graph_lock=graph_lock)
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/graph/graph_manager.py", line 290, in _load_graph
    apply_build_requires=apply_build_requires) 
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/graph/graph_manager.py", line 257, in _recurse_build_requires
    processed_profile, graph_lock)
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/client/graph/graph_builder.py", line 57, in extend_build_requires
    graph_lock.lock_node(node, requires)
  File "/home/tamas/src/repros/conan/build_requires_keyerror/.venv/lib/python3.6/site-packages/conans/model/graph_lock.py", line 252, in lock_node
    locked_pref, locked_id = prefs[require.ref.name]
KeyError: 'fontconfig'
```

---

To provide a little background on why we are using build_requires this way: 

  * Our ffmpeg package contains an executable that is statically linked and used in different places where it's undesirable to download all the other deps at install time (as they are completely unused in those builds)
  * We are cross-compiling ffmpeg and its deps for targeting windows (via mingw running on Linux). This yields packages that have `settings.compiler` set to `gcc`. But on Windows, we are using msvc. Our earlier attempts were at collecting these packages and generating a profile that overrides the compiler for these packages so they can be consumed in an MSVC build, but this is not ideal. 
  * We initially used private requires for this purpose but were advised by your team to use `build_requires` instead due to faulty design in the former.

Our workaround for now is probably going to be creating a facade package that repackages the real ffmpeg package with the requires and settings that we really want to expose. 
   

---

To help us debug your issue please explain:

- [x] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [x] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

**Comment by memsharded** [comment]
Ok, I think I have identified the issue.

When you create the lock, you need to lock the build_requires too. So use:

```bash
conan graph lock . --build=cascade  --build=outdated
```

Otherwise, the build-requires are simply not there, not locked at all, so if after you try to compute the build order asking for them, they will fail to resolve. 

I will try to improve the error message and diagnostics, to help identify this issue, and investigate if there could be a better approach, more robust.

**Comment by lasote** [comment]
The traceback error should be fixed at 1.19.0, now it will show a controlled error, but the root "issue" is still there. We will think about it for 1.20

**Comment by memsharded** [comment]
We haven't come to a solution yet.

So far it seems that the approach would be to provide ``--build`` arguments to ``conan graph lock``, and that would store the nodes to be built, so later ``conan graph build-order`` will use the stored information in the lockfile about the to-be-built nodes. Just an idea, we will need more time to explore it, moving it to next 1.21, still as look-into.

**Comment by memsharded** [comment]
Solved in Conan 2.0, as build_requires are always part of the graph. Please test in the 2.0-beta.1 already available.
